### PR TITLE
New behavior wrt exception

### PR DESCRIPTION
**Pull request by panglesd**
This PR implements the change in behavior proposed in #314.

When a located exception is raised during a rewriting pass, and either the `as-ppx` or the `embed_errors` flag is set, then instead of replacing the ast by an error extension node, the last valid AST (the one before the error) is used and prepended with the error extension node.

### CODE REVIEW

**Review by pitag-ha — COMMENTED** [review]
Nice, thanks @panglesd!

Apart from the few comments below, I just have one observation: you've chosen the approach of lifting the needed information to the entry function `process_ast` via throwing and catching exceptions containing that information along the way. You then transform that information into what we want in `process_ast` depending on the variable `embed_errors`. Another approach would have been to pass down `embed_errors` and to directly transform the information into what we want the first time we catch it (in `apply_transforms`). Have you also considered that?

File: src/driver.ml [hunk 1]
```diff
@@ -900,6 +939,34 @@ struct
   let set x = t.data <- Some x
 end
 
+(*$*)
```
[the remark this answers is not in the document]
Comment (pitag-ha) [reply src/driver.ml:942]: Are you using cinaps for that? If so, why?

File: src/driver.ml [hunk 2]
```diff
@@ -900,6 +939,34 @@ struct
...
+
+let error_to_str_extension error =
+  let loc = Location.none in
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml:945]: Yes, it's because it is the location of the error extension node, not the location of the error itself.
The extension node has been added to the AST after parsing, it has no location in the `.ml` or `.mli` file.

[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml:945]: Yes, it's because it is the location of the error extension node, not the location of the error.
The extension node has been added to the AST after parsing, it has no location in the `.ml` or `.mli` file.

File: src/driver.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml:942]: Yes, cinaps is used to chack that `str` have their exact `sig` counterparts, using [this function](https://github.com/ocaml-ppx/ppxlib/blob/d973c9584dbc5145493bd9bf18f07312596bad5b/src/cinaps/ppxlib_cinaps_helpers.ml#L3).
So if you modify one function but not its counterpart, `dune runtest` will show a diff and offer a promotion.

These special comments separate cinaps blocks.

File: src/driver.ml [hunk 3]
```diff
@@ -556,92 +563,124 @@ let print_passes () =
...
+         checks than for actual rewriting. *)
+      Attribute.check_unused#structure st;
+      if !perform_checks_on_extensions then Extension.check_unused#structure st;
+      Attribute.check_all_seen ();
+      if !perform_locations_check then
+        let open Location_check in
+        ignore
+          ((enforce_invariants !loc_fname)#structure st
+             Non_intersecting_ranges.empty
+            : Non_intersecting_ranges.t));
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml:598]: You are right :dart:.

I also agree that we should embed the generated errors in the AST. I can do that in this PR as well.

File: src/driver.ml [hunk 4]
```diff
@@ -900,6 +939,34 @@ struct
...
+let error_to_str_extension error =
+  let loc = Location.none in
+  let ext = Location.Error.to_extension error in
+  let open Ast_builder.Default in
+  pstr_extension ~loc ext []
+
+let exn_to_str_extension exn =
+  match Location.Error.of_exn exn with
+  | None -> raise exn
+  | Some error -> error_to_str_extension error
```
[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml]: Thanks, I forgot to use them in the "unspecialized" `impl_or_intf` version of these functions.

File: src/driver.ml (hunk 3 above)
[the remark this answers is not in the document]
Comment (panglesd) [reply src/driver.ml:598]: In fact, I think it might be better to do it in another PR, focused on replacing the `raise_errorf` calls by `error_extensionf`, when PR #316 has been merged. (I count 34 calls of `raise_errorf`).

**Comment by panglesd** [comment]
> Another approach would have been to pass down embed_errors and to directly transform the information into what we want the first time we catch it (in apply_transforms). Have you also considered that?

Sure, that is also a sound approach, but I faced technical problems:
We need some kind of `fold_until` function.
Due to the absence of polymorphism in exceptions, and the fact that `apply_transforms` is polymorphic (there is only one version for both `structure` and `signature`), it is not really possible to "simulate" one with `fold_left` and exceptions (apart from using locally abstract types and first class modules, as is done in `Base`...)
So, I decided to catch the exceptions later, at a point where signature vs structure is disambiguated.

However, there is also the possibility to write our own `fold_until` function without exceptions, that I may not have considered (I don't know why).
The fact that we want to catch only the exceptions raised during [linting](https://github.com/ocaml-ppx/ppxlib/pull/315/files#diff-4ccb0a6338feee150dace4d3baa90ef90b2a60453ff5279b1535cbf915652d2eR524) and [transforming](https://github.com/ocaml-ppx/ppxlib/pull/315/files#diff-4ccb0a6338feee150dace4d3baa90ef90b2a60453ff5279b1535cbf915652d2eR531), makes it a little verbose, but it could be improved with some version of `Result.map` (tough `Result` should not be used for compatibility wirh `<4.08`).

In summary, I think that you are right, it is a better solution. The last commit implements this.

File: src/driver.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply src/driver.ml]: Yes. What I was surprised about was to see cinaps used for boilerplate generation of a function of a couple of lines. I've usually seen it used to make the code future-proof. Seems great also to use it for boilerplate generation though, particularly now that you're re-using those functions!

File: src/driver.ml (hunk 2 above)
[the remark this answers is not in the document]
Comment (pitag-ha) [reply src/driver.ml]: > The extension node has been added to the AST after parsing, it has no location in the .ml or .mli file.

Notice that any node added by a ppx rewriter or ppxlib is added to the AST after parsing and still it's important to give those nodes a good location. More concretely, nodes added by the parser have a natural location in the source code file, whereas the locations of nodes added by a ppx rewriter / ppxlib need to be clarified by the author of the rewriter.

Anyways, I've just seen that we were already using `Location.none` in this context before this PR and you've only moved it, so let's keep it like that for now while I'll find out what would be the right thing to do in this context for the future.

File: CHANGES.md [hunk 5]
```diff
@@ -19,6 +19,9 @@
   consider a type declaration recursive if the type appeared inside an attribute
   payload (#299, @NathanReb)
 
+- Append the last valid AST to the error in case of located exception when
+  `embed_errors` is true (#315, @panglesd)
+
```
Comment (pitag-ha) [inline CHANGES.md:24]: Oh, this is on me: I should have created an empty `unreleased` section right after the last release! On your current branch there's no unreleased section and you've added this to the section of the last release. Would you mind rebasing over main and moving this entry to the `unreleased` section?

And just two things about the entry itself: the user doesn't know what our internal reference `embed_errors` is. So instead of saying "when `embed_errors` is true", you could either formulate it in terms of flags "when `-embed-errors` or `-as-ppx` is set" or in terms of situation "when embedding errors".

And the second thing is that it might be nice to give the context that you're talking about the driver, e.g.: 
```
- Driver: append the last valid AST to the error in case of located exception when (...)
```

**Review by pitag-ha — COMMENTED** [review]
Perfect, thanks for the improvements!

> In summary, I think that you are right, it is a better solution.

To be honest, I didn't have an opinion on that, I was really just asking, but both solutions seem good! About the current solution: It's not very clear to me why the `map_<structure, signature>_gen` return type contains information about if the rewriting was successful or not while `process_ast` doesn't, even though currently it isn't needed in any of them. Do you think it's more likely that in the future that information might be needed by functions calling `map_<structure, signature>_gen` than by functions calling `process_ast`? In any case, of course that doubt isn't blocking, so let's merge this as soon as you've moved your changelog entry to `unreleased`!

**Comment by pitag-ha** [comment]
Thanks for moving and improving the changelog entry!

I was just about to merge when noticing the following: shouldn't we add a `merlin.hide` attribute to the structure item with the error extension node?

**Review by pitag-ha — APPROVED** [review]
Perfect, let's merge this. Thanks for your work, @panglesd!

Just for future reference about my comment:

> shouldn't we add a `merlin.hide` attribute to the structure item with the error extension node?

As we've just learned from @voodoos, both the compiler and merlin give error extension nodes a special treatment during the construction of the typedtree: the compiler raises when encountering them and merlin collects those nodes separately from the typedtree. So those nodes don't end up in the typedtree and therefore adding a `merlin.hide` attribute wouldn't have any impact and isn't necessary.
